lib/rpc/SingleJsonRpcProvider.ts

```diff
@@ -296,13 +295,17 @@ export class SingleJsonRpcProvider extends StaticJsonRpcProvider {
       return await fn(...args)
     } catch (error: any) {
       perf.succeed = false
-      this.log.error(JSON.stringify(error))
+      this.log.debug(
+        `Provider call failed: provider: ${this.url}, fnName: ${fnName}, fn: ${fn}, args: ${JSON.stringify([...args])}`
+      )
+      this.log.error(`Error details: ${JSON.stringify(error)}`)
       throw error
     } finally {
       perf.latencyInMs = Date.now() - perf.startTimestampInMs
       this.checkLastCallPerformance(perf)
       this.updateHealthyStatus()
-      if (this.enableDbSync) {
+      if (this.enableDbSync && !this.syncingDb) {
```

Comment:
`enableDbSyn` and `syncingDb` feel like duplicate configs to me. `syncingDb` hardcoded to false feels like a shortcut to me, that may potentially get lost in future.

Reply:
No they are different. `enableDbSync` means whether we enabled DB sync feature at all and it's only set in the constructor. `syncingDb` on the other hand is set to indicate we are in the progress of a DB sync (act like a lock to prevent future sync requests from being executed).

Reply:
it looks like you are trying to implement an optimistic lock with boolean, which probably won't work because your if condition is not atomic. Thoughts on either using an existing library or implementing our own so we can transform this block into the following ? 
```
if (this.enableDbSync){
   this.lock.tryAcquire(this.maybeSyncAndUpdateProviderState)
}
```

IMHO, it would clean up a bit the booleans in this class and provide this primitive for other projects